### DISCUSSION TITLE
How can I prevent `brew install x` from touching anything else?

### DISCUSSION THREAD

**Discussion by dgrcode** [Homebrew/discussions#3305] [Everyday usage]
I'm not sure if this is an issue with my configuration or if I'm missing something, but my experience with brew is far from good.

Yesterday I had an issue with openssl, so decided to install the package to use it. I just did `brew install openssl`. After a few minutes doing some checks, brew decided to update many other packages. I didn't know it yet, but this broke a few programs that relied on a specific version of those packages. It updated stuff for a few more minutes. Then, after 15-20 minutes it told me openssl was already installed.

**First question here. How is it possible that I have to wait ~20 minutes to be told that the package I wanted to install is already installed?**

I went to do whatever I was going to do with openssl, and to my surprise I couldn't because I had the wrong version of mysql. Now I have a new quest where I have to go and fix all the programs that are broken due to wrong version of packages.

**Second question. How is it possible that brew goes and update stuff that no one told it to update?**

I don't know how much time I'll lose just because I tried to install a package that I already had, but I cannot understand how it's possible to end up in this position.

**Third question. Is there a configuration somewhere where I can tell homebrew to do _only_ what I ask it to do?**

Thanks in advance!

**Comment by SMillerDev** [comment]
> but this broke a few programs that relied on a specific version of those packages.

Homebrew is a rolling release package manager so it is bad at specific versions by design

> First question here. How is it possible that I have to wait ~20 minutes to be told that the package I wanted to install is already installed?

Since clearly your install was very far behind the state of the art. Since Homebrew is a rolling release package manager it needed to catch up first.

> Second question. How is it possible that brew goes and update stuff that no one told it to update?

You told it to install OpenSSL, it needed newer dependencies then you had installed so it updated them. In the process it also updated everything that depended on the software it updated, since that would be broken otherwise. The fact that you already had OpenSSL installed was only later realized.

> Third question. Is there a configuration somewhere where I can tell homebrew to do only what I ask it to do?

It did exactly what you asked it to do though. If you mean "can I set up my system in such a way that it will run ancient software until I explicitly update that software, potentially breaking anything I install anew but I will take care of those issues myself" then yes. There is an environment variable for that in https://docs.brew.sh/Manpage

EDIT: https://docs.brew.sh/FAQ#why-does-brew-upgrade-formula-or-brew-install-formula-also-upgrade-a-bunch-of-other-stuff is also related

**Comment by tlk** [comment]
> **Third question. Is there a configuration somewhere where I can tell homebrew to do _only_ what I ask it to do?**

I have found that adding `export HOMEBREW_NO_AUTO_UPDATE=1` to ~/.zshrc makes Homebrew less likely to break things.

This will re-establish the original behaviour where the user is expected to run `brew update` (like `apt update`) instead of the default setting where `brew install foo` runs `brew update` under the hood ➞ which triggers these cascading upgrades ➞ which increases the risk that things will break unexpectedly.
